common/base/src/logging.cpp

```diff
@@ -27,66 +29,67 @@ namespace genomeworks
 
 namespace logging
 {
-static std::shared_ptr<spdlog::logger> logger = nullptr;
+static std::unique_ptr<std::ostream> out_stream_ = nullptr;
 
-LoggingStatus Init(const char* filename)
+static LogLevel level_ = LogLevel::error;
+
+void check_logger()
 {
-    // for now, first call wins:
-    if (logger != nullptr)
-        return LoggingStatus::success;
+    if (out_stream_ == nullptr)
+    {
+        std::cerr << "GenomeWorks logger not initialized yet. Initializing default logger now." << std::endl;
+        initialize_logger(LogLevel::error);
+    }
+}
+
+std::string log_level_str(LogLevel level)
+{
+    std::string prefix;
+    switch (level)
+    {
+    case critical: prefix = "CRITICAL"; break;
+    case error: prefix = "ERROR"; break;
+    case warn: prefix = "WARN"; break;
+    case info: prefix = "INFO"; break;
+    case debug: prefix = "DEBUG"; break;
+    default:
+        assert(false); // Unknown log level
+        prefix = "INFO";
+        break;
+    }
+    return prefix;
+}
 
-    if (filename != nullptr)
+void initialize_logger(LogLevel level, const char* filename)
+{
+    if (out_stream_ == nullptr)
     {
-        try
+        level_ = level;
+        if (filename != nullptr)
         {
-            logger = spdlog::basic_logger_mt("GWLogger", filename);
+            out_stream_ = std::make_unique<std::ofstream>(filename);
         }
-        catch (const spdlog::spdlog_ex& ex)
+        else
         {
-            return LoggingStatus::cannot_open_file;
+            out_stream_ = std::make_unique<std::ostream>(std::cerr.rdbuf());
         }
+        assert(out_stream_->good());
```

Comment:
Here you should throw an exception "Could not open ...", since `out_stream_.good() == false` could happen even if the program itself is correct. It might happen due to a file permission problem, etc.

Reply:
done